app/models/spree/ability.rb

```diff
@@ -29,6 +29,7 @@ def initialize(user)
         can :update, Order do |order, token|
           order.user == user || order.token && token == order.token
         end
+        can [:index, :generate, :create], :invoice
```

Comment:
Oh! We've mixed up CanCan _actions_ and _subjects_. I had to read up on this: https://github.com/CanCanCommunity/cancancan/blob/develop/docs/define_check_abilities.md

In the controller, we check if we can perform action `:invoice` on subject `Order`. But there's no rule for `can :invoice, Order`. Either, we should be checking if we can 
* `:index`, `:generate` or `:create` an `Invoice` (which is hard because these are bulk actions and the invoice doesn't exist yet). OR
* `invoice` an `Order` and update the rules here.

And finally, I think we need to re-think how these rules are applied. This part of of the file is a list of generic rules for all customers and admins; the admin-only section is down below.

Reply:
Hi @dacook 

> In the controller, we check if we can perform action :invoice on subject Order. But there's no rule for can :invoice, Order. 

I think we're already doing that on `add_order_management_abilities`
https://github.com/openfoodfoundation/openfoodnetwork/blob/de34b9b1098a0872f077abdba2995685b8aa1f94/app/models/spree/ability.rb#L266-L274

but after reading your comment and checking the code again, I think that I updated the permissions from the wrong place.  I'll update the permissions `:index, :generate, :create` from `add_order_management_abilities` (instead of the initialize)


Now, when the user tries to generate an invoice, the `authorize!` will be called (at least) three times:

2 times from  `‎Spree::Admin::BaseController`:
1. to check if the user has `admin` permission on `:invoice`
2. to check if the user has `generate` permission on `:invoice`
https://github.com/openfoodfoundation/openfoodnetwork/blob/de34b9b1098a0872f077abdba2995685b8aa1f94/app/controllers/spree/admin/base_controller.rb#L45-L57

3. the last call (from the action method) is to check if the user has `invoice` permission on the `order`.

I don't know if we can make things simpler.